daemon/containerd/image.go

```diff
@@ -2,14 +2,236 @@ package containerd
+		children, err := handler(ctx, desc)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, err
+			}
+		}
+
+		atomic.AddInt64(&size, desc.Size)
+
+		return children, nil
+	}
+
+	l := semaphore.NewWeighted(3)
+	if err := containerdimages.Dispatch(ctx, wh, l, desc); err != nil {
+		return 0, err
+	}
+
+	return size, nil
+}
+
+// resolveDescriptor searches for a descriptor based on the given
+// reference or identifier. Returns the descriptor of
+// the image, which could be a manifest list, manifest, or config.
+func (i *ImageService) resolveDescriptor(ctx context.Context, refOrID string, platform *ocispec.Platform) (ocispec.Descriptor, error) {
+	parsed, err := reference.ParseAnyReference(refOrID)
+	if err != nil {
+		return ocispec.Descriptor{}, errdefs.InvalidParameter(err)
+	}
+
+	is := i.client.ImageService()
+
+	digested, ok := parsed.(reference.Digested)
+	if ok {
+		imgs, err := is.List(ctx, "target.digest=="+digested.Digest().String())
+		if err != nil {
+			return ocispec.Descriptor{}, errors.Wrap(err, "failed to lookup digest")
+		}
+		if len(imgs) == 0 {
+			return ocispec.Descriptor{}, images.ErrImageDoesNotExist{Ref: parsed}
+		}
+
+		return imgs[0].Target, nil
+	}
+
+	ref := reference.TagNameOnly(parsed.(reference.Named)).String()
+
+	// If the identifier could be a short ID, attempt to match
+	if truncatedID.MatchString(refOrID) {
+		filters := []string{
+			fmt.Sprintf("name==%q", ref), // Or it could just look like one.
+			"target.digest~=" + strconv.Quote(fmt.Sprintf(`sha256:^%s[0-9a-fA-F]{%d}$`, regexp.QuoteMeta(refOrID), 64-len(refOrID))),
+		}
+		imgs, err := is.List(ctx, filters...)
+		if err != nil {
+			return ocispec.Descriptor{}, err
+		}
+
+		if len(imgs) == 0 {
+			return ocispec.Descriptor{}, images.ErrImageDoesNotExist{Ref: parsed}
+		}
+		if len(imgs) > 1 {
+			digests := map[digest.Digest]struct{}{}
+			for _, img := range imgs {
+				if img.Name == ref {
+					return img.Target, nil
+				}
+				digests[img.Target.Digest] = struct{}{}
+			}
+
+			if len(digests) > 1 {
+				return ocispec.Descriptor{}, errdefs.NotFound(errors.New("ambiguous reference"))
+			}
+		}
+
+		return imgs[0].Target, nil
+	}
+
+	img, err := is.Get(ctx, ref)
+	if err != nil {
+		// TODO(containerd): error translation can use common function
+		if !cerrdefs.IsNotFound(err) {
+			return ocispec.Descriptor{}, err
+		}
+		return ocispec.Descriptor{}, images.ErrImageDoesNotExist{Ref: parsed}
+	}
+	if platform != nil {
+		cs := i.client.ContentStore()
+		imgPlatforms, err := containerdimages.Platforms(ctx, cs, img.Target)
+		if err != nil {
+			return img.Target, err
+		}
+
+		comparer := platforms.Only(*platform)
+		for _, p := range imgPlatforms {
+			if comparer.Match(p) {
+				return img.Target, nil
+			}
+		}
+		return img.Target, errdefs.NotFound(errors.Errorf("%s: platform %s not supported by image", refOrID, platforms.Format(*platform)))
+	}
```

Comment:
I think this should be extracted from `resolveDescriptor` to `GetImage`. Accepting `platform` as an argument to this function gives an impression that the returned descriptor will be platform specific and it doesn't.
The platform is relevant only in the `GetImage` context, which uses it to pick the correct image config.

Reply:
👍 I moved the code